### PR TITLE
Part 3

### PR DESCRIPTION
**Pull request by benjimarshall**

### CODE REVIEW

File: SupportBank/Transaction.cs [hunk 1]
```diff
@@ -11,13 +11,20 @@ internal class Transaction
...
 
-        public Transaction(DateTime date, Person from, Person to, string narrative, double amount)
+        public Transaction(DateTime date, string fromStr, string toStr, string narrative, double amount,
+                           Dictionary<string, Person> people)
         {
             Date = date;
-            From = from;
-            To = to;
+            From = FindOrAddPerson(fromStr, people);
+            To = FindOrAddPerson(toStr, people);
```
Comment (FridaTveit) [inline SupportBank/Transaction.cs:19]: I'm not a big fan of this happening here, it doesn't seem like this should be the responsibility of the transaction object. What about instead when you have the parsed json transaction list, you iterate through it and for each transaction call some `processTransaction` function which will create the transaction object and update the people dictionary. You could use that function for both reading json and reading csv files, as the function can just take the individual fields as arguments, like this constructor 🙂

**Review by FridaTveit — COMMENTED** [review]
Nice work! I've added some comments but otherwise looks good 🙂

File: SupportBank/CliInterface.cs [hunk 2]
```diff
@@ -61,5 +72,34 @@ private static void PrintBalances(Dictionary<string, Person> people)
         {
             foreach (var person in people.Values) Console.WriteLine(person.BalanceStatus());
         }
+
+        private static void ReadFromFile(string filename, Dictionary<string, Person> people)
+        {
+            var fileTypeMatcher = new Regex(@".*\.((csv)|(json))");
+
+            if (fileTypeMatcher.IsMatch(filename))
```
[the remark this answers is not in the document]
Comment (benjimarshall) [reply SupportBank/CliInterface.cs]: Yep 🙂 Done in 32b01cf

File: SupportBank/JsonReader.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace SupportBank
+{
+    class JsonReader
+    {
+        public static void ReadJson(string filename, Dictionary<string, Person> people)
+        {
+            Program.Logger.Debug($"Starting to parse {filename}");
+
+            using (var file = File.OpenText(filename))
```
[the remark this answers is not in the document]
Comment (benjimarshall) [reply SupportBank/JsonReader.cs]: I didn't know that, thanks! Changed in 32b01cf

File: SupportBank/Transaction.cs [hunk 4]
```diff
@@ -11,13 +11,20 @@ internal class Transaction
         public readonly string Narrative;
         public readonly Person To;
 
-        public Transaction(DateTime date, Person from, Person to, string narrative, double amount)
+        public Transaction(DateTime date, string fromStr, string toStr, string narrative, double amount,
+                           Dictionary<string, Person> people)
```
[the remark this answers is not in the document]
Comment (benjimarshall) [reply SupportBank/Transaction.cs]: Function declaration fixed in 32b01cf. I've ended up refactoring a bunch of this in b924295 so it matches the JSON formatting based on your later comment, so it has the more helpful names `fromName` and `toName` 🙂

File: SupportBank/JsonReader.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+                var transactionList = (List<JsonTransaction>)serializer.Deserialize(file, typeof(List<JsonTransaction>));
+
+                transactionList.Select(jsonTransaction => new Transaction(
+                    jsonTransaction.Date,
+                    jsonTransaction.FromAccount,
+                    jsonTransaction.ToAccount,
+                    jsonTransaction.Narrative,
+                    jsonTransaction.Amount,
+                    people
+                )).ToList();
```
[the remark this answers is not in the document]
Comment (benjimarshall) [reply SupportBank/JsonReader.cs]: Not at the moment, I was trying to make the lazy evaluation work. I've got a better version of this in b924295 using IEnumerable.ForEach()

File: SupportBank/Transaction.cs (hunk 1 above)
Comment (benjimarshall) [reply SupportBank/Transaction.cs]: Yeah you're right. In b924295 I moved wrapped the `people` dictionary in its own object called `AllPeople`, which means there's fewer `Dictionary<string, Person>` types floating around. In this I've added a method to add transactions to the people inside the directory, as well as other operations required for the directory. I think it's now more consistent than it was before and solves the problem you've just described 🙂

**Review by FridaTveit — APPROVED** [review]
Looks good 🙂

**Pull request merged**